### PR TITLE
PostsPage Refactor + ToC updates

### PR DESCRIPTION
**Pull request by Raemon**

### CODE REVIEW

File: packages/lesswrong/components/posts/PostsPage/PostsPageEventData.jsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import { registerComponent, Components } from 'meteor/vulcan:core';
+import { withStyles } from '@material-ui/core/styles'
+import withUser from '../../common/withUser'
+
+const styles = theme => ({
+  metadata: {
+    marginTop:theme.spacing.unit*3,
+    ...theme.typography.postStyle,
+    color: 'rgba(0,0,0,0.5)',
+    fontSize: "1.2rem",
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsPage/PostsPageEventData.jsx]: Hmm, would obviously prefer to keep font-size proliferation to a minimum, but might be inevitable.

File: packages/lesswrong/components/posts/TableOfContents/TableOfContents.jsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  return function WithToCContextComponent(props) {
+    return <TableOfContentsContext.Consumer>
+      { setToC => <Component {...props} setToC={setToC}/> }
+    </TableOfContentsContext.Consumer>
+  }
+}
+
+class TableOfContents extends Component
+{
+  constructor(props) {
```
Comment (Discordius) [inline packages/lesswrong/components/posts/TableOfContents/TableOfContents.jsx]: Remove constructor and put state directly as a class property

File: packages/lesswrong/components/posts/TableOfContents/TableOfContents.jsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    super(props);
+
+    this.state = {
+      drawerOpen: false,
+    };
+  }
+
+  componentDidMount() {
+    this.props.setToC(this.props.document, this.props.sections);
+    // addCallback('router.onUpdate', this.props.setToC(null));
```
Comment (Discordius) [inline packages/lesswrong/components/posts/TableOfContents/TableOfContents.jsx]: Remove comment?

File: packages/lesswrong/components/posts/TableOfContents/TableOfContentsList.jsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import withErrorBoundary from '../../common/withErrorBoundary'
+
+const topSection = "top";
+
+class TableOfContentsList extends Component
+{
+  constructor(props) {
+    super(props);
+
+    this.state = {
```
Comment (Discordius) [inline packages/lesswrong/components/posts/TableOfContents/TableOfContentsList.jsx]: Would also remove constructor here and replace with class property

File: packages/lesswrong/components/posts/TableOfContents/TableOfContentsRow.jsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    ...theme.typography.body2,
+    ...theme.typography.postStyle,
+  },
+
+  // For the highlighted section only, disable the half-opacity-on-hover effect
+  // that's otherwise globally applied to <a> tags.
+  highlighted: {
+    fontWeight: "bold",
+    '& $link': {
+      color: "black !important",
```
Comment (Discordius) [inline packages/lesswrong/components/posts/TableOfContents/TableOfContentsRow.jsx]: I am quite sure this isn't necessary. Though you might have to increase specificity.

File: packages/lesswrong/components/posts/TableOfContents/TableOfContentsRow.jsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        { [classes.highlighted]: highlighted }
+      )}
+    >
+      <a href={href} onClick={onClick} className={classes.link}>
+        {children}
+      </a>
+    </Typography>
+  }
+
+  levelToClassName = (level) => {
```
Comment (Discordius) [inline packages/lesswrong/components/posts/TableOfContents/TableOfContentsRow.jsx]: You have this function twice, maybe we want to abstract it

File: packages/lesswrong/lib/collections/posts/custom_fields.js [hunk 7]
```diff
@@ -1023,6 +1025,23 @@ Posts.addField([
...
+      type: Object,
+      optional: true,
+      viewableBy: ['guests'],
+      resolveAs: {
+        fieldName: "tableOfContents",
+        type: GraphQLJSON,
+        resolver: async (document, args, options) => {
+          return Utils.extractTableOfContents(document.htmlBody);
+        },
+        addOriginalField: true,
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/posts/custom_fields.js]: Why would you want to add the original field here? If I am not confused, there should never be data in there. I would unify the two variable names, and then remove this.

File: packages/lesswrong/components/common/Section.jsx [hunk 8]
```diff
@@ -54,6 +54,9 @@ const styles = (theme) => ({
     maxWidth: 739,
     [theme.breakpoints.up('md')]: {
       paddingLeft: 24
+    },
+    [theme.breakpoints.down('sm')]: {
```
[the remark this answers is not in the document]
Comment (Raemon) [reply packages/lesswrong/components/common/Section.jsx]: Poked around in events, sequences, groups and frontpage. All seem to work fine (or at least, any not-fine-ness was already there in the previous Section version)

File: packages/lesswrong/components/posts/TableOfContents/TableOfContentsRow.jsx (hunk 6 above)
Comment (Raemon) [reply packages/lesswrong/components/posts/TableOfContents/TableOfContentsRow.jsx]: Agreed, although not 100% sure where it should live.

**Comment by Discordius** [comment]
Other comments: 

I quite dislike the new date formatting, and would prefer to either do the same we used to have, or do relative dates on desktop. The new dateformat feels quite ugly to me.

**Comment by Discordius** [comment]
Also, the current date doesn't display the full date with timezone on hover, which I think is bad and we should be consistent for all dates.

File: packages/lesswrong/.babelrc [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "presets": ["env"],
+  "plugins": [
+    [
+      "module-resolver", {
```
[the remark this answers is not in the document]
Comment (jimrandomh) [reply packages/lesswrong/.babelrc]: Doh, this wasn't supposed to be in this commit (it was part of an unrelated thing I was experimenting with in my working directory). Will take it back out.

File: packages/lesswrong/lib/collections/posts/custom_fields.js (hunk 7 above)
Comment (jimrandomh) [reply packages/lesswrong/lib/collections/posts/custom_fields.js]: Me imitating existing resolvers without fully understanding what all the fields do, IIRC.

File: packages/lesswrong/components/posts/PostsPage/PostsPage.jsx [hunk 10]
```diff
@@ -160,31 +180,37 @@ class PostsPage extends Component {
...
-                      <FromNowDate date={post.postedAt}/>
-                    </Hidden>
-                  </span>
+                  {!post.isEvent && <span className={classes.mobileDate}>
+                    <FromNowDate date={post.postedAt}/>
+                  </span>}
+                  {!post.isEvent && <span className={classes.desktopDate}>
+                    {moment(post.postedAt).format("DD MMM YYYY")}
+                  </span>}
+                  {post.types && post.types.length && <Components.GroupLinks document={post} />}
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsPage/PostsPage.jsx]:
```suggestion
                  {post.types && (post.types.length > 0) && <Components.GroupLinks document={post} />}
```

**Pull request merged**